sentry_dsn_data.go

```diff
@@ -0,0 +1,163 @@
+package main
+
+import (
+	"context"
+	"errors"
+	"sync"
+
+	"github.com/getsentry/sentry-go"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+var DSN = "dsn"
+
+// map from Sentry DSN to Client
+type DsnData struct {
+	mutex     sync.RWMutex
+	clientMap map[string]*sentry.Client
+}
+
+func NewDsnData() *DsnData {
+	return &DsnData{
+		mutex:     sync.RWMutex{},
+		clientMap: make(map[string]*sentry.Client),
+	}
+}
+
+// return client if added successfully
+func (d *DsnData) AddClient(options sentry.ClientOptions) (*sentry.Client, error) {
+	d.mutex.Lock()
+	defer d.mutex.Unlock()
+
+	// check if we already encountered this dsn
+	existingClient, ok := d.clientMap[options.Dsn]
+	if ok {
+		return existingClient, errors.New("a client with the given dsn already exists")
+	}
+
+	// create a new client for the dsn
+	newClient, err := sentry.NewClient(
+		sentry.ClientOptions{
+			Dsn:              options.Dsn,
+			Debug:            true,
+			AttachStacktrace: true,
+		},
+	)
+	if err != nil {
+		return nil, err
+	}
+	d.clientMap[options.Dsn] = newClient
+
+	return newClient, nil
+}
+
+// retrieve a client with given dsn
+func (d *DsnData) GetClient(dsn string) (*sentry.Client, error) {
+	d.mutex.RLock()
+	defer d.mutex.RUnlock()
+
+	// check if we have this dsn
+	existingClient, ok := d.clientMap[dsn]
+	if ok {
+		return existingClient, nil
+	} else {
+		return nil, errors.New("a client with given DSN does not exist")
+	}
+}
+
+// recursive function to find if there is a DSN annotation
+func searchDsn(ctx context.Context, object metav1.ObjectMeta) (string, error) {
+
+	clientset, err := getClientsetFromContext(ctx)
+	if err != nil {
+		return "", err
+	}
+
+	dsn, ok := object.Annotations[DSN]
+	if ok {
+		return dsn, nil
+	}
+
+	if len(object.OwnerReferences) == 0 {
+		return "", nil
+	}
+
+	owningRef := object.OwnerReferences[0]
+	switch kind := owningRef.Kind; kind {
```

Comment:
Could you try to move this switch-case to a separate function that e.g. returns a more generic Interface? https://github.com/kubernetes/client-go/blob/84a6fe7e4032ae1b8bc03b5208e771c5f7103549/metadata/interface.go#L34-L43 (in the example of Pods -- the result of `clientset.CoreV1().Pods(object.Namespace)`)

This might help in deduplicating a very similar switch-case from `findObjectMeta`.

Reply:
I just used findObjectMeta as a helper function here and slightly changed the recursive function